### PR TITLE
New CUDA Fuser code lowering refactor

### PR DESCRIPTION
**Pull request by csarofeen**
This PR completely refactors the code lowering process from our IR to CUDA. Before we had one giant step that would go from a relatively high level IR straight to CUDA, now we're lowering this first into concepts like ForLoop, IfThenElse, TensorIndex, Allocate. This lowering will allow us to do more complex code lowering like reductions and unrolling. Unrolling will quickly follow this PR.

### CODE REVIEW

**Comment by jjsjann123** [comment]
Looks like we still have code_write.h included:
`/cpp/jit/test_gpu.cpp:5:10: fatal error: torch/csrc/jit/codegen/cuda/code_write.h:`

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 1]
```diff
@@ -55,6 +161,52 @@ bool IRInputOutput::hasOutput(const Val* const output) const {
...
+      " .");
+}
+
+void IRInputOutput::removeInput(Val* val) {
+  auto it = inputs_.begin();
+  for (; it != inputs_.end(); ++it) {
+    if ((*it) == val)
+      break;
+  }
+  assert(it != inputs_.end());
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:206]: I had few instances where assert is not doing the right thing for me. Maybe we should switch to `TORCH_INTERNAL_ASSERT`

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 2]
```diff
@@ -155,6 +172,111 @@ bool Reorder::sameAs(const Reorder* const other) const {
...
+    if (!(index(i)->sameAs(other->index(i))))
+      return false;
+
+  return true;
+}
+
+Val* TensorIndex::index(int i) const {
+  if (i < 0)
+    i += size();
+  assert(i >= 0 && i < size());
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp:242]: same here.

**Review by jjsjann123 — APPROVED** [review]
LGTM.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: changed

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp:242]: changed

**Review by soumith — APPROVED** [review]

**Pull request closed without merging**